src/main/java/org/rumbledb/compiler/StaticContextVisitor.java

```diff
@@ -102,18 +112,31 @@ public StaticContext visitFunctionDeclaration(FunctionDeclaration declaration, S
         InlineFunctionExpression expression = (InlineFunctionExpression) declaration.getExpression();
         // define a static context for the function body, add params to the context and visit the body expression
         StaticContext functionDeclarationContext = new StaticContext(argument);
-        expression.getParams()
-            .forEach(
-                (paramName, sequenceType) -> functionDeclarationContext.addVariable(
-                    paramName,
-                    sequenceType,
-                    expression.getMetadata(),
-                    ExecutionMode.LOCAL // static udf currently supports materialized(local) params, not RDDs or DFs
-                )
+        List<ExecutionMode> modes = Functions.getUserDefinedFunctionParametersStorageMode(
+            expression.getFunctionIdentifier(),
+            expression.getMetadata()
+        );
+        int i = 0;
+        for (String name : expression.getParams().keySet()) {
+            ExecutionMode mode = modes.get(i);
+            SequenceType type = expression.getParams().get(name);
+            if (type.isEmptySequence()) {
+                mode = ExecutionMode.LOCAL;
+            } else if (type.getArity().equals(Arity.OneOrZero) || type.getArity().equals(Arity.One)) {
+                mode = ExecutionMode.LOCAL;
+            }
+            functionDeclarationContext.addVariable(
+                name,
+                expression.getParams().get(name),
+                expression.getMetadata(),
+                mode
             );
+            ++i;
+        }
```

Comment:
Is it possible to extract this code block to a function?

Reply:
Good idea, this is now separated.